### PR TITLE
Yedit

### PR DESCRIPTION
**Pull request by elharo**
@briandealwis I'm trying to add yedit to the plugin, but the target platform isn't finding in it. Any ideas what I might be doing wrong?

### CODE REVIEW

**Comment by briandealwis** [comment]
Hmm, it's working for me: I see yedit in `ide-target-platform` and in `gcp-repo`.
```sh
$ (cd eclipse; mvn package)
…
[INFO] BUILD SUCCESS
…
$ ls eclipse/ide-target-platform/target/repository/plugins/*yedit*
eclipse/ide-target-platform/target/repository/plugins/org.dadacoalition.yedit_1.0.18.201602092025-RELEASE-SIGNED.jar

$ mvn package
…
[INFO] BUILD SUCCESS
…
$ ls gcp-repo/target/repository/plugins/*yedit*
gcp-repo/target/repository/plugins/org.dadacoalition.yedit_1.0.18.201602092025-RELEASE-SIGNED.jar
```

**Comment by briandealwis** [comment]
Do we want to point to http://dadacoalition.org/yedit/ instead? (It's an update site.)

**Comment by elharo** [comment]
And yet it doesn't seem to show up in Eclipse:

image

**Comment by elharo** [comment]
Description	Resource	Path	Location	Type
Plug-in reference 'org.dadacoalition.yedit' cannot be resolved	feature.xml	/com.google.cloud.tools.eclipse.3rdparty.feature	line 85	Plug-in Problem

**Comment by briandealwis** [comment]
You've done the [https://github.com/GoogleCloudPlatform/google-cloud-eclipse#updating-dependencies reload dance], right?

**Comment by elharo** [comment]
yep, repeatedly. And the new appengine-plugins-core did get picked up.

**Comment by briandealwis** [comment]
Weird: now I'm seeing this too.

On reflection, there's an easier way for us to set up our `ide-target-platform`; I'll push up a better approach in a sec.

**Comment by elharo** [comment]
Finally got this to work by using the new repo loction and 0.0.0 version.

At the moment, this is a proof of concept that this can work. We need to discuss seriously whether we should bundle this or not.

**Comment by elharo** [comment]
FYI 

https://github.com/oyse/yedit/issues/36

**Comment by elharo** [comment]
I want to move this forward but once again I'm having problems getting Eclipse to pick up yedit. The plugin jar shows up in the repo but the feature doesn't get installed.

**Comment by elharo** [comment]
It's in the target platform:

image

**Comment by elharo** [comment]
After a lot of reloading and rebuilding I got it to work again:

image

I'm a little concerned this does not appear to have generated any actual code changes.

**Comment by elharo** [comment]
Can you elaborate on the why you're not pulling down the org.dadacoalition.yedit.feature feature? What do you expect to see here that you don't?

File: eclipse/mars/gcp-eclipse-mars.tpd [hunk 1]
```diff
@@ -59,3 +59,7 @@ location "http://download.eclipse.org/tools/orbit/downloads/drops/R2016022119215
     ch.qos.logback.slf4j
     org.slf4j.log4j
 }
+
+location "http://dadacoalition.org/yedit/" {
+  org.dadacoalition.yedit
```
Comment (briandealwis) [inline eclipse/mars/gcp-eclipse-mars.tpd]: Pull down `org.dadacoalition.yedit.feature.feature.group`

File: eclipse/neon/gcp-eclipse-neon.tpd [hunk 2]
```diff
@@ -55,3 +55,7 @@ location "http://download.eclipse.org/tools/orbit/downloads/drops/R2016052021185
     ch.qos.logback.slf4j
     org.slf4j.log4j
 }
+
+location "http://dadacoalition.org/yedit/" {
+  org.dadacoalition.yedit
```
Comment (briandealwis) [inline eclipse/neon/gcp-eclipse-neon.tpd]: Ditto

File: eclipse/oxygen/gcp-eclipse-oxygen.tpd [hunk 3]
```diff
@@ -55,3 +55,7 @@ location "http://download.eclipse.org/tools/orbit/downloads/drops/R2017051619251
     ch.qos.logback.slf4j
     org.slf4j.log4j
 }
+
+location "http://dadacoalition.org/yedit/" {
+  org.dadacoalition.yedit
```
Comment (briandealwis) [inline eclipse/oxygen/gcp-eclipse-oxygen.tpd]: Ditto.

File: features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml [hunk 4]
```diff
@@ -78,7 +78,14 @@
          id="org.yaml.snakeyaml"
          download-size="0"
          install-size="0"
-         version="1.17"
+         version="1.17.0"
+         unpack="false"/>
+         
+ <plugin
+         id="org.dadacoalition.yedit"
```
Comment (briandealwis) [inline features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: Should use the import-feature to pull in `org.dadacoaltion.yedit.feature`.

File: eclipse/ide-target-platform/category.xml [hunk 5]
```diff
@@ -37,6 +37,7 @@
    <bundle id="com.fasterxml.jackson.core.jackson-core" version="0.0.0"/>
    <bundle id="ch.qos.logback.slf4j" version="0.0.0"/>
    <bundle id="org.slf4j.log4j" version="0.0.0"/>
+   <bundle id="org.dadacoalition.yedit" version="0.0.0"/>
```
Comment (briandealwis) [inline eclipse/ide-target-platform/category.xml]: bring in the `org.dadacoalition.yedit.feature` feature instead

File: eclipse/mars/gcp-eclipse-mars.tpd (hunk 1 above)
Comment (elharo) [reply eclipse/mars/gcp-eclipse-mars.tpd]: What do you mean by "pull down"? Change org.dadacoalition.yedit to org.dadacoalition.yedit.feature.feature.group on line 61?

File: features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml (hunk 4 above)
Comment (chanseokoh) [reply features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: What is the import-feature? How does that work?

Comment (briandealwis) [reply features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: We can either use the _Included Features_ tab (which binds to a specific version, which is ok in our case), or the _Dependencies_ tab which allows importing a feature or plugin with a version range.

Comment (elharo) [reply features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: It sounds like you're referring to something in the PDE UI. I'm not really using that. The question is what the code should say and what it means.

Comment (elharo) [reply features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: I think what you're asking for is something like this:

```
  <requires>
    <import
         feature="org.dadacoalition.yedit.feature"
         version="0.0.0"/>
  </requires>
```

However, "Feature reference 'org.dadacoalition.yedit.feature' cannot be resolved	feature.xml	/com.google.cloud.tools.eclipse.3rdparty.feature	line 100"

Comment (elharo) [reply features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: Or maybe this?

```
  <requires>
    <import plugin="org.dadacoalition.yedit"/>
  </requires>
```

Comment (briandealwis) [reply features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: The _Included Features_ is:
```
   <includes
         id="org.dadacoalition.yedit.feature"
         version="0.0.0"/>
```

**Comment by chanseokoh** [comment]
Dismissed the approval to double-check. Currently the PR stands as
```xml
<requires>
  <import plugin="org.dadacoalition.yedit"/>
</requires>
```
but I guess [@briandealwis suggested `<import feature="...">`](https://github.com/GoogleCloudPlatform/google-cloud-eclipse/pull/2016#pullrequestreview-73210545).

**Comment by chanseokoh** [comment]
How about `org.dadacoalition.yedit.feature.feature.group`? Worth a try.

**Comment by elharo** [comment]
Nope:

```
Feature reference 'org.dadacoalition.yedit.feature.feature.group' cannot be resolved	feature.xml	/com.google.cloud.tools.eclipse.3rdparty.feature	line 99	Plug-in Problem
```

**Comment by elharo** [comment]
Worth noting: if I use the PDE feature editor, it cannot find any feature for org.dadacoalition though it can find the plugin org.dadacoalition.yedit. Perhaps some other change must be made first, maybe in target platform?

**Comment by briandealwis** [comment]
You did update your target platform prior, right?  Building from the command-line works for me, and I see the YEdit plugin and feature under `eclipse/ide-target-platform/target/repository/`:
```
$ ls -l eclipse/ide-target-platform/target/repository/*/*yedit*
-rw-r--r--  1 bsd  staff    6267 31 Oct 14:23 eclipse/ide-target-platform/target/repository/features/org.dadacoalition.yedit.feature_1.0.20.201509041456-RELEASE.jar
-rw-r—r--  1 bsd  staff  537320 31 Oct 14:23 eclipse/ide-target-platform/target/repository/plugins/org.dadacoalition.yedit_1.0.20.201509041456-RELEASE.jar
```

The following patch can be either applied to `.3rdparty.feature` or `.suite.e45.feature`; YEdit won'tbe installed if we put it in `.3rdparty.feature` since we have nothing that references it.  We could expose it to be installable though by referencing it from the `gcp-repo/metadata.p2.inf` (or create and reference it from a `gcp-repo/category.xml`).
```diff
--- a/feature.xml
+++ b/feature.xml
@@ -18,6 +18,10 @@
       %license.text
    </license>
 
+   <includes
+      id="org.dadacoalition.yedit.feature"
+      version="0.0.0"/>
+
    <plugin
          id="com.google.cloud.tools.appengine"
          download-size="0"
```

**Comment by elharo** [comment]
OK, finally got Eclipse to recognize the feature. PTAL.

File: features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml [hunk 6]
```diff
@@ -18,6 +18,16 @@
       %license.text
    </license>
 
+   <includes
+         id="org.dadacoalition.yedit.feature"
+         version="0.0.0"
+         name="YEdit"
+         optional="true"/>
+
+   <requires>
```
Comment (briandealwis) [inline features/com.google.cloud.tools.eclipse.3rdparty.feature/feature.xml]: Why both the `<includes>` and `<requires>`? Why optional?

**Comment by elharo** [comment]
Includes and requires because PDE put them both in there. 

Optional in case the update site vanishes. I don't want CT4E to not be installable if YEdit can't be found.

**Comment by elharo** [comment]
Should we add it to the .suite.e45 feature instead?

**Comment by briandealwis** [comment]
Maybe once we have a definite need for it?  The `optional` flag tells p2 to not fail if the dependency isn't found, but since it will be found (since it's in our gcp-repo) it will always be installed.

**Pull request merged**